File: frontend/src/app/features/tarjetas/tarjetas.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { environment } from '../../../environments/environment';

export interface TarjetaOut {
  tarjeta_id: number;
  cuenta_id: number;
  tipo_tarjeta: 'debito' | 'credito';
  ultimos_4: string;
  fecha_emision: string;
  fecha_vencimiento: string;
  estado: string;
}

/** Solo existe en la respuesta de /revelar: nunca se persiste ni se guarda en un servicio compartido. */
export interface TarjetaRevelada {
  pan: string;
  vencimiento: string;
  cvv: string;
  titular: string;
}

@Injectable({ providedIn: 'root' })
export class Tarjetas {
  private readonly http = inject(HttpClient);

  listar() {
    return this.http.get<TarjetaOut[]>(`${environment.apiUrl}/tarjetas`);
  }

  revelar(tarjetaId: number, password: string) {
    return this.http.post<TarjetaRevelada>(`${environment.apiUrl}/tarjetas/${tarjetaId}/revelar`, { password });
  }
}

File: backend/app/schemas/tarjetas.py
from datetime import date

from pydantic import BaseModel, ConfigDict, Field


class TarjetaOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    tarjeta_id: int
    cuenta_id: int
    tipo_tarjeta: str
    ultimos_4: str
    fecha_emision: date
    fecha_vencimiento: date
    estado: str


class RevelarIn(BaseModel):
    password: str = Field(min_length=1)


class TarjetaRevelada(BaseModel):
    """CVV dinamico: cambia en cada revelado, no se almacena en ninguna tabla (PCI-DSS 3.3.1)."""

    pan: str
    vencimiento: str  # "MM/AA"
    cvv: str
    titular: str
